junit5/src/main/java/cz/xtf/junit5/listeners/OpenshiftStateRecorder.java

```diff
@@ -0,0 +1,199 @@
+
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.LocalDateTime;
+
+import static org.junit.platform.engine.TestExecutionResult.Status.FAILED;
+
+/**
+ * Listener for storing information from openshift in case a test fails.
+ * It stores pod logs, openshift routes, secret, events etc. From both work and build namespace.
+ * It implements two distinct interfaces TestExecutionListener and LifecycleMethodExecutionExceptionHandler
+ * both serving different purpose and need to be applied differently.
+ *
+ * TestExecutionListener
+ *    Is called whenever test ends. Stores openshift state if test failed.
+ *    To use this interface add this class as a listener
+ *
+ * LifecycleMethodExecutionExceptionHandler
+ *    This interface is executed when @BeforeAll or @BeforeEach method fails - throw an exception.
+ *    To use this interface a test class needs to be extended with this class using @ExtendWith
+ */
+@Slf4j
+public class OpenshiftStateRecorder implements TestExecutionListener, LifecycleMethodExecutionExceptionHandler {
+	private static final Path statusDir = Paths.get("log","status");
+
+	private final OpenShift openShift = OpenShifts.master();
+	private final OpenShift buildOpenShift;
+
+	private final String namespace;
+	private final String buildNamespace;
+
+	public OpenshiftStateRecorder(){
+		namespace = openShift.getNamespace();
+		buildNamespace = XTFConfig.get(BuildManagerConfig.BUILD_NAMESPACE);
+		buildOpenShift = OpenShifts.master(buildNamespace);
+	}
+
+	private String getTestDisplayName(TestIdentifier testIdentifier) {
+		String className = testIdentifier.getParentId().get()
+			.replaceAll(".*class:", "")
+			.replaceAll("].*", "");
+		return String.format("%s#%s", className, testIdentifier.getDisplayName());
+	}
+
+	// Log the openshift state if BeforeAll fail
+	@Override
+	public void handleBeforeAllMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if BeforeEach fail
+	@Override
+	public void handleBeforeEachMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if test fail
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		try {
+			if (FAILED.equals(testExecutionResult.getStatus()) && testIdentifier.isTest()) {
```

Comment:
Could that be configurable, so I am able to store the state also for PASSED test, e.g. with something like `xtf.config.store_state_always` ?

By default it will store infor just for failed tests.